File: bend2/effs/tcp_listen.js
// TCP
// ===
//! use ./sys.js

function tcp_listen(port) {
  const sys = sys_get();
  const fd = sys.sock(sys.SOCK_STREAM);
  if (fd < 0) {
    return sys.fail(sys.errno());
  }
  sys.reuse(fd);
  const at = sys.addr("0.0.0.0", Number(port));
  if (at === null) {
    sys.s.close(fd);
    return sys.fail(22);
  }
  if (sys.s.bind(fd, sys.ptr(at), 16) < 0 || sys.s.listen(fd, 16) < 0) {
    const code = sys.errno();
    sys.s.close(fd);
    return sys.fail(code);
  }
  return sys.done(sys.mint("Listener", "lsn", fd));
}

File: bend2/effs/tcp_listen.c
// TCP
// ===

Term tcp_listen_run(Env e, Term* f, IoWork* w) {
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) {
    return io_fail(e, (uint32_t)errno, NULL);
  }
  int one = 1;
  setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
  struct sockaddr_in at;
  if (io_sys_addr("0.0.0.0", (uint32_t)f[0], &at) < 0) {
    close(fd);
    return io_fail(e, EINVAL, NULL);
  }
  int bound = bind(fd, (struct sockaddr*)&at, sizeof(at));
  if (bound < 0 || listen(fd, 16) < 0
    || fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    uint32_t code = (uint32_t)errno;
    close(fd);
    return io_fail(e, code, NULL);
  }
  return io_done(e, io_hand(fd));
}

static void __attribute__((constructor)) tcp_listen_use(void) {
  io_eff(CID(TCP.listen), tcp_listen_run, 0);
}
